File: src/main/java/core/procedures/generic/Reverse.java
package core.procedures.generic;

import core.exceptions.WrongTypeException;
import core.procedures.AFn;
import core.procedures.FnArgsBuilder;
import core.scm.*;
import core.scm.Vector;

import java.util.*;

public final class Reverse extends AFn {

  public Reverse() {
    super(new FnArgsBuilder().min(1).max(1).build());
  }

  @Override
  public String getName() {
    return "reverse";
  }

  @Override
  public Object apply1(Object arg) {
    if (arg instanceof List) {
      Cons<Object> result = Cons.list();
      for (Object o : (List) arg) {
        result.push(o);
      }
      return result;
    }
    if (arg instanceof Set) {
      return Cons.list((Set)arg);
    }
    if (arg instanceof Map.Entry) {
      return new MapEntry(((Map.Entry) arg).getValue(), ((Map.Entry) arg).getKey());
    }
    if (arg instanceof Vector) {
      Object[] array = ((Vector) arg).getArray();
      MutableVector reversed = new MutableVector(Arrays.copyOf(array, array.length));
      Collections.reverse(Arrays.asList(reversed.getArray()));
      return reversed;
    }
    if (arg instanceof CharSequence) {
      return new StringBuilder((CharSequence) arg).reverse().toString();
    }
    throw new WrongTypeException(getName(), "List or Vector or Set or String", arg);
  }
}

File: src/main/java/core/procedures/generic/Reverse.kt
package core.procedures.generic

import core.exceptions.WrongTypeException
import core.procedures.AFn
import core.scm.Cons
import core.scm.MapEntry
import core.scm.MutableVector
import core.scm.Vector

class Reverse : AFn<Any?, Any?>(name = "reverse", isPure = true, minArgs = 1, maxArgs = 1) {

    override operator fun invoke(arg: Any?): Any? = when (arg) {
        is List<*>         -> Cons.list<Any>().apply { for (o in (arg as List<*>?)!!) { add(0, o) } }
        is Set<*>          -> Cons.list(arg as Collection<Any?>)
        is Map.Entry<*, *> -> MapEntry(arg.value, arg.key)
        is CharSequence    -> StringBuilder((arg as CharSequence?)!!).reverse().toString()
        is ByteArray       -> arg.copyOf().apply { reverse() }
        is Vector -> {
            val array = arg.getArray()
            MutableVector(array.size, null).apply {
                for (i in array.indices) {
                    this.array[i] = array[array.size - i - 1]
                }
            }
        }
        else -> throw WrongTypeException(name, "List or Vector or Set or String", arg)
    }
}
